HW1/P5/P5_bfs.py

```diff
@@ -0,0 +1,74 @@
+# P5 method
+
+def ssbfs(nodestart, directedlist):
+    sc = directedlist.context
+    directedlist.cache()
+    visited = set([nodestart])  # visited nodes
+    current = set([nodestart])  # nodes of the current level
+    result = sc.parallelize([(nodestart, None)])  # track the path, child-parent pairs
+
+    while True:
+        if len(current) == 0:  # if no node in current level
+            break  # end the bfs
+
+        waitinglist = directedlist.filter(lambda kv: kv[0] in current)  # reach out from current level
+        waitinglist = waitinglist.flatMapValues(lambda v: v)
+        waitinglist = waitinglist.map(lambda kv: (kv[1], kv[0]))
+        nextlist = waitinglist.filter(lambda kv: not kv[0] in visited)  # new nodes we found
+        nextlist = nextlist.reduceByKey(lambda a, b: a)  # choose one parent is enough
+        result = result.union(nextlist)  # update our result
+        
+        current = set(nextlist.keys().collect())  # update two sets for next iteration
+        visited = visited.union(current)
+    return result
+
+def shortestpath(nodestart, nodeend, directedlist):
+    sc = directedlist.context
+    directedlist.cache()
+    visited = set([nodestart])  # visited nodes
+    current = set([nodestart])  # nodes of the current level
+    result = sc.parallelize([(nodestart, None)])  # track the path, child-parent pairs
+
+    while True:
+        if nodeend in current:  # if found
+            path = [nodeend]  # trace back from nodeend
+            while True:
+                pathway = result.lookup(path[-1])[0]
+                if pathway:
+                    path.append(pathway)
+                else:
+                    break
+            return len(path) - 1, path[::-1]
+
+        if len(current) == 0:  # if no node in current level
+            break  # end the bfs
+
+        waitinglist = directedlist.filter(lambda kv: kv[0] in current)  # reach out from current level
+        waitinglist = waitinglist.flatMapValues(lambda v: v)
+        waitinglist = waitinglist.map(lambda kv: (kv[1], kv[0]))
+        nextlist = waitinglist.filter(lambda kv: not kv[0] in visited)  # new nodes we found
```

Comment:
Could this also be done with a flag in each element of the RDD to specify whether the node was unseen, about to be processed, or processed? This way it should also save on the expense of calls to collect() and then reinsertion.